### PR TITLE
feature: support icon graphs

### PR DESCRIPTION
**Pull request by fprill**
This PR introduces some data structures that define a processor mesh and an encoder/decoder graph based on the icosahedral ICON grid, read from a NetCDF grid  file.

----
📚 Documentation preview 📚: https://anemoi-graphs--53.org.readthedocs.build/en/53/

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=53) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=53) before we can accept your contribution.<br/><hr/>**Florian Prill** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-graphs?pullRequest=53) it.</sub>

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+       Lam, R. et al. (2022). GraphCast: Learning skillful medium-range
+       global weather forecasting. 10.48550/arXiv.2212.12794.
+
+    """
+
+    phi_theta = gc_recv  # precession and nutation angle
+    theta__phi = np.fliplr(phi_theta)
+    theta__phi[:, 1] *= -1.0  # angles = [theta, -phi]
+    # rotation matrix:
+    R = scipy.spatial.transform.Rotation.from_euler(seq="YZ", angles=theta__phi)
```
Comment (JesperDramsch) [inline src/anemoi/graphs/generate/icon_mesh.py:470]: nit: not a fan of single-letter capital variable names

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+    Returns:
+      boolean matrix given by list of column indices for each row.
+    """
+    if remove_duplicates:
+        # The edges-vertex adjacency matrix may have duplicate rows, remove
+        # them by selecting the rows that are unique:
+        nrows = int(adj_matrix.nnz / ncols_per_row)
+        mat = adj_matrix.indices.reshape((nrows, ncols_per_row))
+        return np.unique(mat, axis=0)
+    else:
```
Comment (JesperDramsch) [inline src/anemoi/graphs/generate/icon_mesh.py:536]: Doesn't need `else` due to the return statement

File: src/anemoi/graphs/nodes/builder.py [hunk 3]
```diff
@@ -282,6 +282,77 @@ def create_nodes(self) -> np.ndarray:
...
+        key corresponding to the ICON mesh (cells and vertices).
+    """
+
+    def __init__(self, name: str, icon_mesh: str) -> None:
+        self.icon_mesh = icon_mesh
+        super().__init__(name)
+
+    def update_graph(self, graph: HeteroData, attr_config: DotDict | None = None) -> HeteroData:
+        """Update the graph with new nodes."""
+        self.multi_mesh = graph[self.icon_mesh]["_multi_mesh"]
```
Comment (JesperDramsch) [inline src/anemoi/graphs/nodes/builder.py:325]: I wonder if this would be possible to factor out similarly to the Edges

**Review by JesperDramsch — COMMENTED** [review]
Hi @fprill, thank you so much for the contribution. It was very nice to read.

I have some comments on variable names for easier reading of the code (especially the big processing sequence), a technicality on the formatting of docstrings and I think two bigger comments.

One is on the iverbosity key, which I believe could be replaced/avoided.
The other is on potentially factoring the edge builders as they seem very similar in kind.

Again thanks for the great readability of the code, comments and style. (Didn't know about `typeguard` yet, I have to read up more on that. Seems quite useful.)

I marked some comments with `nit` as they're more nitpicks than actual blockers. Additionally, I think some of the larger comments can definitely be seen as a discussion basis rather than a blocker, let me know what you think.

Edit: I haven't left comments on a lot of nice sections in the code, as I didn't want to clutter the review even more. But just wanted to note that I did feel like there are a lot of really nice sections and it was a great read.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+    # -> convert to regular arrays
+    assert not arr.mask.any(), f"There are missing values in {arrname}"
+    return arr.data
+
+
+@typechecked
+def convert_list_to_adjacency_matrix(list_matrix: np.ndarray, ncols: int = 0) -> scipy.sparse.csr_matrix:
+    """Convert an edge list into an adjacency matrix.
+
+    Args:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 69ae2fa84e88f9ec86381edb52064708dcf2748a.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+          number of nonzero entries per row
+      remove_duplicates: bool
+          logical flag: remove duplicate rows.
+    Returns:
+      boolean matrix given by list of column indices for each row.
+    """
+    if remove_duplicates:
+        # The edges-vertex adjacency matrix may have duplicate rows, remove
+        # them by selecting the rows that are unique:
+        nrows = int(adj_matrix.nnz / ncols_per_row)
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 5e21ff5ce9309130ecb339ad7ee3f7115cfb7fcc.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+    def __init__(self, nodeset: NodeSet, bidirectional: bool, edge_vertices: np.ndarray):
+        self.nodeset = nodeset
+        # (optional) duplicate edges (bi-directional):
+        if bidirectional:
+            self.edge_vertices = np.concatenate([edge_vertices, np.fliplr(edge_vertices)])
+        else:
+            self.edge_vertices = edge_vertices
+
+    @property
+    def num_verts(self) -> int:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit d858d0e1298772e0b46b9f10fdccc809eb9f6c1f.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+    theta__phi[:, 1] *= -1.0  # angles = [theta, -phi]
+    # rotation matrix:
+    R = scipy.spatial.transform.Rotation.from_euler(seq="YZ", angles=theta__phi)
+    edge_length = np.array([arc_length(cc_recv, cc_send)])
+    distance = R.apply(cc_send) - [1.0, 0.0, 0.0]  # subtract the rotated position of sender
+    return np.concatenate((edge_length.transpose(), distance), axis=1)
+
+
+@typechecked
+def arc_length(v1, v2) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit 6f6e50143818b25ecd1681a0f82eabd23b292320.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+        nrows = int(adj_matrix.nnz / ncols_per_row)
+        mat = adj_matrix.indices.reshape((nrows, ncols_per_row))
+        return np.unique(mat, axis=0)
+    else:
+        nrows = adj_matrix.shape[0]
+        return adj_matrix.indices.reshape((nrows, ncols_per_row))
+
+
+@typechecked
+def selection_matrix(idx: np.ndarray, isize: int = 0) -> scipy.sparse.csr_matrix:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 0fae14d06573d0c195ecab0c8db5a87b1d350fcb.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+
+@typechecked
+@dataclass
+class EdgeID:
+    """Stores additional categorical data for each edge (IDs for heterogeneous input)."""
+
+    edge_id: np.ndarray
+    num_classes: int
+
+    def __add__(self, other: Self):
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py:74]: I thought that the `typing_extensions` library should work for Python versions prior to 3.9, including older versions like 3.7 and 3.8

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+
+    uuidOfHGrid: str
+    max_level: int
+    nodeset: NodeSet  # set of ICON grid vertices
+    cell_vertices: np.ndarray
+
+    def __init__(self, icon_grid_filename: str, max_level: Optional[int] = None, iverbosity: int = 0):
+
+        # open file, representing the finest level
+        if iverbosity > 0:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: Good point! This has been handled by the commit 368427ca0356d82117592f19c05f77ce2a0b2905.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+        # cell-to-vertex adjacency matrix with 3 non-zero entries per row:
+        c2v_matrix = convert_list_to_adjacency_matrix(cell_verts_fine, num_verts)
+        v2v_matrix = e2v_matrix.transpose() * e2v_matrix
+        v2v_matrix.setdiag(np.ones(num_verts))  # vertices are self-connected
+
+        s_v_coarse = scipy.sparse.diags(np.ones(num_verts), dtype=bool)
+
+        # coarsen edge-vertex list from level `ilevel -> ilevel - 1`:
+        for ilevel in reversed(range(1, reflvl_vertex.max() + 1)):
+            if iverbosity > 0:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 368427ca0356d82117592f19c05f77ce2a0b2905.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+
+    def __init__(
+        self,
+        icon_grid_filename: str,
+        multi_mesh: Optional[ICONMultiMesh] = None,
+        max_level: Optional[int] = None,
+        iverbosity: int = 0,
+    ):
+        # open file, representing the finest level
+        if iverbosity > 0:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 368427ca0356d82117592f19c05f77ce2a0b2905.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 13]
```diff
@@ -0,0 +1,553 @@
...
+      node lies at latitude 0.
+
+    Literature: Appdx. A2.4 in
+       Lam, R. et al. (2022). GraphCast: Learning skillful medium-range
+       global weather forecasting. 10.48550/arXiv.2212.12794.
+
+    """
+
+    phi_theta = gc_recv  # precession and nutation angle
+    theta__phi = np.fliplr(phi_theta)
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 48230ec1c197fd1d7ffb4883c19ad3c98d621d74.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 14]
```diff
@@ -0,0 +1,553 @@
...
+        edge_vertices = np.stack((src_list, dst_list), axis=1, dtype=np.int64)
+        return edge_vertices
+
+
+# -------------------------------------------------------------
+
+
+@typechecked
+def get_icon_mesh_and_grid(
+    iverbosity: int, grid_file: str, max_level_multimesh: int, max_level_dataset: int
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 368427c.

File: src/anemoi/graphs/generate/icon_mesh.py (hunk 1 above)
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit 6ad883f073da15d9a70ac80d2c670dfeb454523e.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 15]
```diff
@@ -0,0 +1,553 @@
...
+    """Graph defined on a pair of NodeSets."""
+
+    nodeset: tuple[NodeSet, NodeSet]  # source and target node set
+    edge_vertices: np.ndarray  # vertex indices for each edge, shape [:,2]
+    edge_id: np.ndarray  # additional ID for each edge (markers for heterogeneous input)
+
+    def __init__(self, nodeset: tuple[NodeSet, NodeSet], edge_vertices: np.ndarray, edge_id: Optional[EdgeID] = None):
+        self.nodeset = nodeset
+        self.edge_vertices = edge_vertices
+        if edge_id is None:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled by the commit a44fb01b38bd83b1151e709b1b46c38f660f8f50.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 16]
```diff
@@ -0,0 +1,553 @@
...
+                self.nodeset[0].gc_vertices[self.edge_vertices[:, 0]],
+                self.nodeset[0].cc_vertices[self.edge_vertices[:, 0]],
+                self.nodeset[1].gc_vertices[self.edge_vertices[:, 1]],
+                self.nodeset[1].cc_vertices[self.edge_vertices[:, 1]],
+            ),
+            dtype=torch.float32,
+        )
+        if self.edge_id is None:
+            return edge_feature_arr
+        else:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit 3d662d171e5f5f13e113b0bc469c4362624836d3.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 17]
```diff
@@ -0,0 +1,553 @@
...
+        """Concatenates two bipartite graphs that share a common target node set.
+        Shifts the node indices of the second bipartite graph.
+        """
+
+        if not self.nodeset[1] == other.nodeset[1]:
+            raise ValueError("Only bipartite graphs with common target node set can be merged.")
+        shifted_edge_vertices = other.edge_vertices
+        shifted_edge_vertices[:, 0] += self.nodeset[0].num_verts
+        # (Optional:) merge one-hot-encoded categorical data (`edge_id`)
+        if None not in (self.edge_id, other.edge_id):
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This  has been handled in the commit 25630cca7056cd86badbf7a538e6d6de5e573a91.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 18]
```diff
@@ -0,0 +1,553 @@
...
+
+            cell_verts_fine = np.asarray(ncfile.variables["vertex_of_cell"][:] - 1, dtype=np.int64).transpose()
+            assert ncfile.variables["vertex_of_cell"].dimensions == ("nv", "cell")
+
+            reflvl_vertex = ncfile.variables["refinement_level_v"][:]
+            assert ncfile.variables["refinement_level_v"].dimensions == ("vertex",)
+
+            self.uuidOfHGrid = ncfile.uuidOfHGrid
+
+        if max_level is not None:
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit 612afe5a68ded4364a435e5f3ffb32dfafa85817.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 19]
```diff
@@ -0,0 +1,553 @@
...
+        edge_vertices: list[np.ndarray],
+        cell_vertices: np.ndarray,
+        reflvl_vertex: np.ndarray,
+        vlon: np.ndarray,
+        vlat: np.ndarray,
+    ) -> tuple[list[np.ndarray], np.ndarray, np.ndarray, np.ndarray]:
+        """Creates a new mesh with only the vertices at the desired level."""
+
+        num_verts = reflvl_vertex.shape[0]
+        vertex_idx = reflvl_vertex <= self.max_level
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit 5c76e10c5f573572b1aa4bab33ad858304930920.

File: src/anemoi/graphs/generate/icon_mesh.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit cfd82a9ebca14336428cddf03294e1d2d836a698.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 20]
```diff
@@ -0,0 +1,553 @@
...
+        vertex_glb2loc = np.zeros(num_verts, dtype=int)
+        vertex_glb2loc[vertex_idx] = np.arange(vertex_idx.sum())
+        return (
+            [vertex_glb2loc[M] for M in edge_vertices[: self.max_level + 1]],
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit cfd82a9ebca14336428cddf03294e1d2d836a698.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 21]
```diff
@@ -0,0 +1,553 @@
...
+        for ilevel in reversed(range(1, reflvl_vertex.max() + 1)):
+            if iverbosity > 0:
+                LOGGER.info(f"  edges[{ilevel}] = {edge_vertices[0].shape[0] : >9}")
+
+            # define edge selection matrix (selecting only edges of which have
+            # exactly one coarse vertex):
+            #
+            # get a boolean mask, matching all edges where one of its vertices
+            # has refinement level index `ilevel`:
+            ref_level = reflvl_vertex[edge_vertices[0]] == ilevel
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This has been handled in the commit f6d63f0056376a93f585ce565bc22360185750d2.

File: src/anemoi/graphs/nodes/builder.py (hunk 3 above)
Comment (fprill) [reply src/anemoi/graphs/nodes/builder.py]: Unfortunately, this doesn't work quite as elegantly for this module as it does for the edges, but I gave it a try in the commit 8efd2c6975163104bc71b4e7e2576a8166a88b7e.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 22]
```diff
@@ -0,0 +1,553 @@
...
+            edges_coarse = np.logical_xor(ref_level[:, 0], ref_level[:, 1])  # = bisected coarse edges
+            idx_e2e = np.argwhere(edges_coarse).flatten()
+            s_edges = selection_matrix(idx_e2e, edges_coarse.shape[0])
+
+            # define vertex selection matrix selecting only vertices of
+            # level `ilevel`:
+            idx_v_fine = np.argwhere(reflvl_vertex == ilevel).flatten()
+            s_v_fine = selection_matrix(idx_v_fine, num_verts)
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This is a valid point indeed. I tried to handle this in the commit 1961150cfc22e62f71a4ab2e26cea7552ed5d79d.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 23]
```diff
@@ -0,0 +1,553 @@
...
+            return edge_feature_arr
+        else:
+            # one-hot-encoded categorical data (`edge_id`)
+            one_hot = torch_geometric.utils.one_hot(
+                index=torch.tensor(self.edge_id.edge_id), num_classes=self.edge_id.num_classes
+            )
+            return torch.concatenate((one_hot, edge_feature_arr), dim=1)
+
+    def set_constant_edge_id(self, edge_id: int, num_classes: int):
+        self.edge_id = EdgeID(edge_id=np.full(self.num_edges, fill_value=edge_id), num_classes=num_classes)
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: This function has become obsolete, yes. I removed it in the commit 30dee38eea8d74924562606bd2d38c77ea9dd3e3.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 24]
```diff
@@ -0,0 +1,553 @@
...
+        return self._gc_to_cartesian()
+
+    def __add__(self, other: Self) -> Self:
+        """concatenates two node sets."""
+        gc_vertices = np.concatenate((self.gc_vertices, other.gc_vertices))
+        return NodeSet(gc_vertices[:, 0], gc_vertices[:, 1])
+
+    def __eq__(self, other: Self) -> bool:
+        """Compares two node sets."""
+        return self.id == other.id
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py:54]: You can't really see this in the code here, but I think that the comparison between two nodesets should be very lightweight and quick to perform: The defining arrays of a nodeset are float arrays with possibly many millions of entries. A real fingerprint would simply be too time-consuming here, I think, so I wanted to rely on comparing the IDs...

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 25]
```diff
@@ -0,0 +1,553 @@
...
+@typechecked
+class NodeSet:
+    """Stores nodes on the unit sphere."""
+
+    id_iter: int = itertools.count()  # unique ID for each object
+    gc_vertices: np.ndarray  # geographical (lat/lon) coordinates [rad], shape [:,2]
+
+    def __init__(self, lon: np.ndarray, lat: np.ndarray):
+        self.gc_vertices = np.column_stack((lon, lat))
+        self.id = next(self.id_iter)
```
[the remark this answers is not in the document]
Comment (fprill) [reply src/anemoi/graphs/generate/icon_mesh.py]: The basic idea behind this design was to differentiate the data locations (nodeset) from the edges in the graphs. In this way, for example, the processor mesh can be defined on a nodeset, but the same nodeset can also represent the target set of the encoder.
However, it is definitely true that object instances should have unique IDs, so the UUID is what we really need here.  I replaced the current code by a `uuid` in the commit 92cb023c2b4d27ef1db83eb76330dd99152c58e0.

File: src/anemoi/graphs/generate/icon_mesh.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JesperDramsch) [reply src/anemoi/graphs/generate/icon_mesh.py:75]: Oh perfect, thanks.

**Comment by HCookie** [comment]
@JPXKQX @gmertes 
Can you take a look at this PR, and approve it if all good?

**Review by JPXKQX — COMMENTED** [review]
Hi @fprill, first of all thanks for the contributions and sorry for the late review. I leave here more general comments: 

- In the last few days we have refactored a few things that may benefit this PR. We have now split the `node/builders.py` into different files under `nodes/builders/`, I think the icon nodes could go into their own ....py file. Would it be interesting to do the same with the edges?

- We also have a PR open for moving some optimising edge attributes builders. I would say it would be beneficial to move the edge_attrs defined in icon_mesh.py to edges/attributes.py. I think for edge direction you can rely on the already implemented and move the edge_length to a new `ArcLength()`. In PR #54 , we have already seen huge performance improvements and in the future we plan to go further to allow parallelisation with different devices. The differences are a few seconds now, but it could play a more important role in higher resolutions.

- Other useful improvements would be to add a test for node & edge builders to avoid breaking functionality with other changes in the future without noticing and to add a `docs/graphs/node_coordinates/icon_mesh.rst` page with the information you consider of interest.

Thanks again for all the work, and I am free to chat to move the PR forward.

**Comment by MeraX** [comment]
@HCookie, I've implemented tests for the ICON grid. Please have a look and approve the workflow to check that they are actually green.

**Comment by MeraX** [comment]
@JPXKQX could you approve again? I had to fix the change log. And could you also trigger the downstream-ci to actually execute pytest?

**Comment by MeraX** [comment]
Perhaps my bad.
https://github.com/ecmwf/anemoi-graphs/actions/runs/11782100162 shows pytest results.
Unfortunately, @fprill was not aware that his code has to support python 3.9.

@JPXKQX do you have a python 3.9 machine at hand? Could you help us find an equivalent syntax that works with Python 3.9? You may push your fixes to our branch.

File: src/anemoi/graphs/generate/icon_mesh.py [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+            num_classes=self.num_classes,
+        )
+
+
+@typechecked
+class GeneralGraph:
+    """Stores edges for a given node set."""
+
+    nodeset: NodeSet  # graph nodes
+    edge_vertices: np.ndarray  # vertex indices for each edge, shape [:,2]
```
Comment (theissenhelen) [inline src/anemoi/graphs/generate/icon_mesh.py:91]: It might be more consistent to call them "edge_indices" as this is the terminology used in the other edge building components.

Comment (MeraX) [reply src/anemoi/graphs/generate/icon_mesh.py:91]: Hi Helen, thanks for your review. Would you agree to keep this refactoring for a follow-up PR? I quickly grepped through `anemoi-*` and could not find anything for `edge_indices`. 
Just in [PyG](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.EdgeIndex.html#torch_geometric.EdgeIndex) I found a definition of `EdgeIndex`. However they  transposed the index order. I think that aligning the index order would definitely beyond the scope of this PR.

However, if you prefer using `edge_indices`, I'd be more than happy to grant you permission to make the necessary changes to this branch. Please do feel free and refactor as needed to get this PR over the line.

**Review by JPXKQX — APPROVED** [review]
Thank you for your contribution!
